# Kids First DRC Consensus Calling Workflow
This workflow is used by the Kids First (KF) Data Resource Center (DRC) to create consensus calls from outputs generated by our somatic variant callers.

![data service logo](https://github.com/d3b-center/d3b-research-workflows/raw/master/doc/kfdrc-logo-sm.png)

This workflow takes the protected vcf outputs from the [Kids First DRC Somatic Workflow](workflow/kfdrc-somatic-variant-workflow.cwl) and creates protected and public consensus VCF and MAF files.
The general outline is as follows:

1. Prep MNP Variants
   - Strelka2 outputs multi-nucleotide polymorphisms (MNPs) as consecutive single-nucleotide polymorphisms
   - In order preserve MNPs, we gather MNP calls from the other caller inputs, and search for evidence supporting these consecutive SNP calls as MNP candidates
    - Once found, the Strelka2 SNP calls supporting a MNP are converted to a single MNP call
    - This is done to preserve the predicted gene model as accurately as possible in our consensus calls
1. Consensus merge
   - Calls are gathered from all four callers
   - By default, calls with support from 2+ callers OR calls that are marked as `HotSpotAllele` in the `INFO` field are retained
   - Retained calls then have their `MQ` and `MQ0` values calculated from the input tumor cram
   - `GT` fields are estimated as "majority rules," and when no majority exists, set as `0/1` by default
   - `AD`, `DP`, and `AF` are calculated as the average value between callers
   - `ADR`, `DPR`, and `AFR` fields are added as the range of values from the previous point, to give the observer a sense on confidence in the value
1. VEP Annotate Consensus (see [Kids First DRC Somatic Variant Annotation Workflow](https://github.com/kids-first/kf-somatic-workflow/blob/master/docs/kfdrc_annotation_wf.md) for details )
1. BCF tools annotate
   - Additional annotation is performed augment VEP annotation
   - While VEP does have extensive gnomad allele frequency annotation, it is limited to exome values. The added gnomad AF only resource we use augments this as an additional `INFO/AF` field to add WGS frequencies
1. Soft filter variants
   - A soft filter is added based on criteria provided
   - By default, we perform soft filtering as outlined in the [KFDRC Annotation Subworkflow](kfdrc_annotation_subworkflow.md#workflow_description_and_kf_recommended_inputs)
1. VCF2MAF protected
   - Here, for convenience of analysis we convert the resultant, soft-filtered VCF (AKA, "Protected VCF") into MAF format
1. Hard filter VCF
   - The Protected VCF is hard filtered on `PASS` and `HotSpotAllele` for reasons outlined in the `Soft filter variants` step
   - This VCF is known as the "Public VCF"
1. VCF2MAF public
1. Rename outputs

## Workflow Description and KF Recommended Inputs

### General workflow inputs, all file references can be obtained [here](https://cavatica.sbgenomics.com/u/kfdrc-harmonization/kf-references/):
- indexed_reference_fasta: Homo_sapiens_assembly38.fasta
- strelka2_vcf
- mutect2_vcf
- lancet_vcf
- vardict_vcf
- cram #Tumor cram recommended for MQ score calculation
- input_tumor_name
- input_normal_name
- output_basename
- tool_name: "consensus_somatic"
- ncallers: # Optional number of callers required for consensus, recommend `2`
- consensus_ram: `3`
- annotation_vcf: gnomad_3.1.1.vwb_subset.vcf.gz # population stats VCF for public filtering
- vep_cache: homo_sapiens_merged_vep_105_indexed_GRCh38.tar.gz
- annot_columns: "INFO/gnomad_3_1_1_AC:=INFO/AC,INFO/gnomad_3_1_1_AN:=INFO/AN,INFO/gnomad_3_1_1_AF:=INFO/AF,INFO/gnomad_3_1_1_nhomalt:=INFO/nhomalt,INFO/gnomad_3_1_1_AC_popmax:=INFO/AC_popmax,INFO/gnomad_3_1_1_AN_popmax:=INFO/AN_popmax,INFO/gnomad_3_1_1_AF_popmax:=INFO/AF_popmax,INFO/gnomad_3_1_1_nhomalt_popmax:=INFO/nhomalt_popmax,INFO/gnomad_3_1_1_AC_controls_and_biobanks:=INFO/AC_controls_and_biobanks,INFO/gnomad_3_1_1_AN_controls_and_biobanks:=INFO/AN_controls_and_biobanks,INFO/gnomad_3_1_1_AF_controls_and_biobanks:=INFO/AF_controls_and_biobanks,INFO/gnomad_3_1_1_AF_non_cancer:=INFO/AF_non_cancer,INFO/gnomad_3_1_1_primate_ai_score:=INFO/primate_ai_score,INFO/gnomad_3_1_1_splice_ai_consequence:=INFO/splice_ai_consequence"
- gatk_filter_name: `[NORM_DP_LOW, GNOMAD_AF_HIGH]`
- gatk_filter_expression: `[ vc.getGenotype('`_insert_norm_sample_id_here_`').getDP() <= 7,gnomad_3_1_1_AF > 0.001 ]`
- bcftools_public_filter: `FILTER="PASS"|INFO/HotSpotAllele=1`
retain_info: "gnomad_3_1_1_AC,gnomad_3_1_1_AN,gnomad_3_1_1_AF,gnomad_3_1_1_nhomalt,gnomad_3_1_1_AC_popmax,gnomad_3_1_1_AN_popmax,gnomad_3_1_1_AF_popmax,gnomad_3_1_1_nhomalt_popmax,gnomad_3_1_1_AC_controls_and_biobanks,gnomad_3_1_1_AN_controls_and_biobanks,gnomad_3_1_1_AF_controls_and_biobanks,gnomad_3_1_1_AF_non_cancer,gnomad_3_1_1_primate_ai_score,gnomad_3_1_1_splice_ai_consequence,MQ,MQ0,CAL,HotSpotAllele"
- retain_fmt: # csv string with FORMAT fields that you want to keep
- retain_ann: "HGVSg"
- maf_center: "."

## Workflow outputs
- `annotated_protected_outputs`: Array of files containing MAF format of PASS hits, `PASS` VCF with annotation pipeline soft `FILTER`-added values, and VCF index
- `annotated_public_outputs`: Same as above, except MAF and VCF have had entries with soft `FILTER` values removed
